File: scripts/make_circular_favicon.js
#!/usr/bin/env node
/**
 * Convert favicon from square to circle and regenerate all sizes.
 */
const fs = require('fs');
const path = require('path');

async function makeCircularFavicon() {
  let sharp;
  try {
    sharp = require('sharp');
  } catch (e) {
    console.error('Error: sharp module not found. Installing...');
    console.error('Please run: npm install sharp');
    process.exit(1);
  }

  const staticDir = path.join(__dirname, '..', 'static');
  const faviconPath = path.join(staticDir, 'favicon.png');

  if (!fs.existsSync(faviconPath)) {
    console.error(`Error: ${faviconPath} not found`);
    process.exit(1);
  }

  // Read the original image
  const image = sharp(faviconPath);
  const metadata = await image.metadata();
  const size = Math.min(metadata.width, metadata.height);

  // Create a circular mask
  const svgMask = `
    <svg width="${size}" height="${size}">
      <circle cx="${size / 2}" cy="${size / 2}" r="${size / 2}" fill="white"/>
    </svg>
  `;

  // Apply circular mask to create circular favicon
  const circularBuffer = await image
    .resize(size, size, { fit: 'cover' })
    .composite([{
      input: Buffer.from(svgMask),
      blend: 'dest-in'
    }])
    .png()
    .toBuffer();

  // Save circular version
  const circularPath = path.join(staticDir, 'favicon_circular.png');
  fs.writeFileSync(circularPath, circularBuffer);
  console.log('Created circular favicon base');

  // Sizes to generate
  const sizes = [
    { width: 180, height: 180, name: 'apple-touch-icon.png' },
    { width: 192, height: 192, name: 'android-chrome-192x192.png' },
    { width: 512, height: 512, name: 'android-chrome-512x512.png' },
    { width: 32, height: 32, name: 'favicon-32x32.png' },
    { width: 16, height: 16, name: 'favicon-16x16.png' },
  ];

  // Generate all sizes
  for (const { width, height, name } of sizes) {
    const svgMaskSize = `
      <svg width="${width}" height="${height}">
        <circle cx="${width / 2}" cy="${height / 2}" r="${width / 2}" fill="white"/>
      </svg>
    `;

    const outputBuffer = await sharp(circularPath)
      .resize(width, height, { fit: 'cover' })
      .composite([{
        input: Buffer.from(svgMaskSize),
        blend: 'dest-in'
      }])
      .png()
      .toBuffer();

    const outputPath = path.join(staticDir, name);
    fs.writeFileSync(outputPath, outputBuffer);
    console.log(`Generated ${name} (${width}x${height})`);
  }

  // Replace original with circular version
  fs.renameSync(circularPath, faviconPath);
  console.log('\n✅ All favicon files have been converted to circular!');
}

makeCircularFavicon().catch(err => {
  console.error('Error:', err.message);
  process.exit(1);
});

File: scripts/make_circular_favicon.py
#!/usr/bin/env python3
"""
Convert favicon from square to circle and regenerate all sizes.
"""
from PIL import Image, ImageDraw
import os
import sys

def make_circular(image_path, output_path):
    """Convert a square image to a circular one."""
    # Open the image
    img = Image.open(image_path).convert("RGBA")
    
    # Get dimensions
    width, height = img.size
    size = min(width, height)
    
    # Create a circular mask
    mask = Image.new("L", (size, size), 0)
    draw = ImageDraw.Draw(mask)
    draw.ellipse((0, 0, size, size), fill=255)
    
    # Resize image to square if needed
    if width != height:
        img = img.resize((size, size), Image.Resampling.LANCZOS)
    
    # Create output image with transparency
    output = Image.new("RGBA", (size, size), (0, 0, 0, 0))
    
    # Apply circular mask
    output.paste(img, (0, 0))
    output.putalpha(mask)
    
    # Save the result
    output.save(output_path, "PNG", optimize=True)
    print(f"Created circular favicon: {output_path}")

def main():
    script_dir = os.path.dirname(os.path.abspath(__file__))
    project_root = os.path.dirname(script_dir)
    static_dir = os.path.join(project_root, "static")
    
    favicon_path = os.path.join(static_dir, "favicon.png")
    
    if not os.path.exists(favicon_path):
        print(f"Error: {favicon_path} not found")
        sys.exit(1)
    
    # Create circular version of the original
    circular_favicon = os.path.join(static_dir, "favicon_circular.png")
    make_circular(favicon_path, circular_favicon)
    
    # Sizes to generate
    sizes = [
        (180, 180, "apple-touch-icon.png"),
        (192, 192, "android-chrome-192x192.png"),
        (512, 512, "android-chrome-512x512.png"),
        (32, 32, "favicon-32x32.png"),
        (16, 16, "favicon-16x16.png"),
    ]
    
    # Generate all sizes
    for width, height, filename in sizes:
        output_path = os.path.join(static_dir, filename)
        make_circular(circular_favicon, output_path)
        # Resize to specific dimensions
        img = Image.open(output_path)
        img = img.resize((width, height), Image.Resampling.LANCZOS)
        img.save(output_path, "PNG", optimize=True)
        print(f"Generated {filename} ({width}x{height})")
    
    # Replace original with circular version
    os.replace(circular_favicon, favicon_path)
    print("\n✅ All favicon files have been converted to circular!")

if __name__ == "__main__":
    main()
